### PR TITLE
rust OpenID Token plugin example

### PR DESCRIPTION
**Pull request by Ross-ForAllSecure**

### CODE REVIEW

**Comment by swarkentin** [comment]
Do you want to add a link to the new example in the plugin folder readme?

https://github.com/ForAllSecure/mapi-examples/blob/main/plugins/README.md

File: plugins/rust-openid-token-plugin/README.md [hunk 1]
```diff
@@ -0,0 +1,135 @@
+This rust plugin example implements an authorization header 
+retreived from an [OAuth OpenID Connect access token endpoint.](https://openid.net/specs/openid-connect-core-1_0.html#rfc.section.3.1.3)
+
+![Sequence Diagram](./sequence_diagram.png)
```
Comment (swarkentin) [inline plugins/rust-openid-token-plugin/README.md:4]: 👍  This diagram is great!!

File: plugins/rust-openid-token-plugin/src/main.rs [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+    oauth_token_url: Url,
+
+    /// This option provides the body of the POST request sent to the `/token`
+    /// endpoint in order to fetch an access token. The data will be urlencoded.
+    /// Multiple calls will be joined with the `&` symbol and the full body of
+    /// the HTTP request will be sent with a `content-type` of
+    /// `application/x-www-form-urlencoded` This switch is a minimalist copy of
+    /// curl's `switch of the same name.
+    #[arg(long = "data-urlencode", env, value_parser(try_parse_data_urlencode))]
+    oath_data_urlencodes: Vec<String>,
```
Comment (swarkentin) [inline plugins/rust-openid-token-plugin/src/main.rs]:
```suggestion
    oauth_data_urlencodes: Vec<String>,
```

Note -- accepting as is will break usage!

File: plugins/rust-openid-token-plugin/src/main.rs [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+
+    /// A header to be sent to on the  oauth enpoint
```
Comment (swarkentin) [inline plugins/rust-openid-token-plugin/src/main.rs]:
```suggestion
    /// A header to be sent to the oauth enpoint
```

File: plugins/rust-openid-token-plugin/src/main.rs [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+    #[arg(
+        long = "header",
+        env = "MAPI_OAUTH_REQUEST_HEADER",
+        value_parser(try_parse_header)
+    )]
+    oauth_request_headers: Vec<(HeaderName, HeaderValue)>,
+
+    /// The name of the HTTP header which will have the access token attached
+    /// when reqeusts are sent to the API under test
```
Comment (swarkentin) [inline plugins/rust-openid-token-plugin/src/main.rs]:
```suggestion
    /// when requests are sent to the API under test
```

File: plugins/rust-openid-token-plugin/src/main.rs [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+        default_value = "authorization",
+        env,
+        value_parser(try_parse_header_name)
+    )]
+    api_under_test_header_name: HeaderName,
+
+    /// A prefix to prepended to the access token and set as the value of the
+    /// header field from `api_under_test_header_name`
+    #[arg(long = "api_under_test_header_prefix", env)]
+    api_under_test_header_prefix: Option<HeaderValue>,
```
Comment (swarkentin) [inline plugins/rust-openid-token-plugin/src/main.rs:61]: It is probably save to default this to `"Bearer "` -- what do you think?

Comment (Ross-ForAllSecure) [reply plugins/rust-openid-token-plugin/src/main.rs:61]: I had it like that, but then I have actually seen headers like `authorization: <jwt>` in the wild and not infrequently.  Then users have to do `--api-under-test-header-prefix=""`, which is rather unwieldy.  I agree that the default would serve more people, but those needing no prefix would find it a little odd.

File: plugins/rust-openid-token-plugin/src/main.rs (hunk 2 above)
Comment (Ross-ForAllSecure) [reply plugins/rust-openid-token-plugin/src/main.rs]: fixed. thanks.

File: plugins/rust-openid-token-plugin/src/main.rs (hunk 5 above)
Comment (swarkentin) [reply plugins/rust-openid-token-plugin/src/main.rs:61]: Ok, that makes sense. I think I would include the `"Bearer "` in the example in the readme in that case (and maybe in the docs here too). Then users can at least see where it is commonly used, without needing to explicitly clear out the default value.

Comment (Ross-ForAllSecure) [reply plugins/rust-openid-token-plugin/src/main.rs:61]: WIll do.

Comment (Ross-ForAllSecure) [reply plugins/rust-openid-token-plugin/src/main.rs:61]: [Done](https://github.com/ForAllSecure/mapi-examples/pull/4/files#diff-83ccb9c9f30f957a078ff94afd053fa8cbccdc063bf5bfbfb4c40f774b37bc3bR90)

**Pull request merged**